inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/sidebar/LearningCurveChartPanel.java

```diff
@@ -18,173 +18,147 @@
-        for (String key : recommenderScoreMap.keySet()) {
-            String data = recommenderScoreMap.get(key).stream().map(Object::toString).collect(Collectors.joining(", "));
+        Map<String,String> curveData = new HashMap<String,String>();
+        LearningCurve learningCurve = new LearningCurve();
+
+        // iterate over recommenderScoreMap to create data
+        for (String recommenderName : recommenderScoreMap.keySet()) {
+            // extract the scores from the recommenderScoreMap. The format of data is a comma
+            // separated string of scores(each score is Double cast-able) to be. 
+            // Example 2.3, 4.5 ,6, 5, 3, 9,
+            String data = recommenderScoreMap.get(recommenderName).stream().map(Object::toString)
+                    .collect(Collectors.joining(", "));
+            
+            curveData.put(recommenderName,data);
             
-            // append recommender name to the data
-            dataColumns.append("['");
-            String[] recommenderClass = key.toString().split("\\.");
-            String recommenderName = recommenderClass[recommenderClass.length - 1];
+            learningCurve.setCurveData(curveData);
+            
+            // the Curve is not allowed to have more points as compared to MAX_POINTS_TO_PLOT. This
+            // is how many scores we have retrieved from the database
+            int[] intArray = IntStream.range(0, MAX_POINTS_TO_PLOT).map(i -> i).toArray();
+            String xaxisValues =  substring(Arrays.toString(intArray), 1, -1)  ;
+            
+            learningCurve.setXaxis(xaxisValues);
+        }
 
-            // define chart type for the recommender
-            chartType.append("'");
-            chartType.append(recommenderName);
-            chartType.append("': 'step', ");
-            dataColumns.append(recommenderName);
+        //provide the chart above calculated data to plot the learning curve
+        chartPanel.setDefaultModel(Model.of(learningCurve));
 
-            // append data columns
-            dataColumns.append("', ");
-            dataColumns.append(data);
-            dataColumns.append("]");
-            dataColumns.append(",");
-        }
- 
+        // to avoid the error, A partial update of the page is being rendered
         try {
-            String javascript = createJSScript(dataColumns.toString(), chartType.toString());
-            log.debug("Rendering Recommender Evaluation Chart: {}", javascript);
-
-            aEvent.getRequestHandler().prependJavaScript(javascript);
+            aRequestHandler.add(chartPanel);
         }
-        catch (IOException e) {
-            log.error("Unable to render chart", e);
-            error("Unable to render chart: " + e.getMessage());
-            aEvent.getRequestHandler().addChildren(getPage(), IFeedback.class);
+        catch (IllegalStateException e) {
+            LOG.warn("Not updating the chart. " + e.toString());
+            setResponsePage(getPage());
```

Comment:
What is the purpose of `setResponsePage(getPage());` here?

Reply:
On the annotation page, recommendation sidebar, clicking on the recommendation sidebar icon again, throws a partial update exception and the chart disappears (Probably because the chart component is already there). `setResponsePage(getPage());` is added in the catch block to refresh the page in order to render the chart again.

Reply:
I'm afraid that is not a good solution. Reloading the page (in particular the annotation page) causes the position of scrollbars to reset which is pretty annoying to the users. What exception is this? Can you provide an error message and a trace?

Reply:
when recommendation side bar icon is clicked. On the line
`            aRequestHandler.add(chartPanel);`
Stack Trace.
```
java.lang.IllegalStateException: A partial update of the page is being rendered, component [ChartPanel [Component id = chart-container]] can no longer be added
	at org.apache.wicket.page.PartialPageUpdate.add(PartialPageUpdate.java:466)
	at org.apache.wicket.ajax.AjaxRequestHandler.add(AjaxRequestHandler.java:241)
	at org.apache.wicket.ajax.AjaxRequestHandler.add(AjaxRequestHandler.java:234)
	at de.tudarmstadt.ukp.inception.recommendation.sidebar.LearningCurveChartPanel.renderChart(LearningCurveChartPanel.java:146)
	at de.tudarmstadt.ukp.inception.recommendation.sidebar.LearningCurveChartPanel.onRender(LearningCurveChartPanel.java:93)
	at org.apache.wicket.Component.internalRender(Component.java:2289)
	at org.apache.wicket.Component.render(Component.java:2220)
	at org.apache.wicket.MarkupContainer.renderNext(MarkupContainer.java:1454)
	at org.apache.wicket.MarkupContainer.renderAll(MarkupContainer.java:1654)
	at org.apache.wicket.MarkupContainer.renderComponentTagBody(MarkupContainer.java:1629)
	at org.apache.wicket.MarkupContainer.renderAssociatedMarkup(MarkupContainer.java:798)
	at org.apache.wicket.markup.html.panel.AssociatedMarkupSourcingStrategy.renderAssociatedMarkup(AssociatedMarkupSourcingStrategy.java:77)
	at org.apache.wicket.markup.html.panel.PanelMarkupSourcingStrategy.onComponentTagBody(PanelMarkupSourcingStrategy.java:112)
	at org.apache.wicket.Component.internalRenderComponent(Component.java:2484)
	at org.apache.wicket.MarkupContainer.onRender(MarkupContainer.java:1593)
	at org.apache.wicket.Component.internalRender(Component.java:2289)
	at org.apache.wicket.Component.render(Component.java:2220)
	at org.apache.wicket.MarkupContainer.renderNext(MarkupContainer.java:1454)
	at org.apache.wicket.MarkupContainer.renderAll(MarkupContainer.java:1654)
	at org.apache.wicket.MarkupContainer.renderComponentTagBody(MarkupContainer.java:1629)
	at org.apache.wicket.MarkupContainer.renderAssociatedMarkup(MarkupContainer.java:798)
	at org.apache.wicket.markup.html.panel.AssociatedMarkupSourcingStrategy.renderAssociatedMarkup(AssociatedMarkupSourcingStrategy.java:77)
	at org.apache.wicket.markup.html.panel.PanelMarkupSourcingStrategy.onComponentTagBody(PanelMarkupSourcingStrategy.java:112)
	at org.apache.wicket.Component.internalRenderComponent(Component.java:2484)
	at org.apache.wicket.MarkupContainer.onRender(MarkupContainer.java:1593)
	at org.apache.wicket.Component.internalRender(Component.java:2289)
	at org.apache.wicket.Component.render(Component.java:2220)
	at org.apache.wicket.Component.renderPart(Component.java:2197)
	at org.apache.wicket.page.XmlPartialPageUpdate.writeComponent(XmlPartialPageUpdate.java:75)
	at org.apache.wicket.page.PartialPageUpdate.writeComponents(PartialPageUpdate.java:262)
	at org.apache.wicket.page.PartialPageUpdate.writeTo(PartialPageUpdate.java:161)
	at org.apache.wicket.ajax.AjaxRequestHandler.respond(AjaxRequestHandler.java:361)
	at org.apache.wicket.request.cycle.RequestCycle$HandlerExecutor.respond(RequestCycle.java:914)
	at org.apache.wicket.request.RequestHandlerExecutor.execute(RequestHandlerExecutor.java:65)
	at org.apache.wicket.request.cycle.RequestCycle.execute(RequestCycle.java:282)
	at org.apache.wicket.request.cycle.RequestCycle.processRequest(RequestCycle.java:253)
	at org.apache.wicket.request.cycle.RequestCycle.processRequestAndDetach(RequestCycle.java:221)
	at org.apache.wicket.protocol.http.WicketFilter.processRequestCycle(WicketFilter.java:275)
	at org.apache.wicket.protocol.http.WicketFilter.processRequest(WicketFilter.java:206)
	at org.apache.wicket.protocol.http.WicketFilter.doFilter(WicketFilter.java:299)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.orm.jpa.support.OpenEntityManagerInViewFilter.doFilterInternal(OpenEntityManagerInViewFilter.java:185)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at de.tudarmstadt.ukp.clarin.webanno.support.logging.LoggingFilter.doFilter(LoggingFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:317)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:114)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:137)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:111)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:158)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:170)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:63)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:66)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:331)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:214)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:177)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:357)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:270)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:112)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:198)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:493)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:800)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:806)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1498)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)
```

Reply:
OnBeforeRender() doesn't remove the exception but the chart doesn't disappear and no need to refresh the page. I removed the problematic line of code that was refreshing the whole page resetting annotation scrollbars. Ideally there needs no refreshing on clicking the sidebar icon because the chart refreshes on render annotation.